### PR TITLE
AP_Filesystem: add support for downloading raw storage via ftp sysfs

### PR DESCRIPTION
**Pull request by bugobliterator**

### CODE REVIEW

File: libraries/AP_Filesystem/AP_Filesystem_Sys.cpp [hunk 1]
```diff
@@ -158,8 +165,13 @@ int32_t AP_Filesystem_Sys::read(int fd, void *buf, uint32_t count)
         return -1;
     }
     struct rfile &r = file[fd];
-    count = MIN(count, r.str->get_length() - r.file_ofs);
-    memcpy(buf, &r.str->get_string()[r.file_ofs], count);
+    if (r.is_storage_bin) {
+        count = MIN(count, HAL_STORAGE_SIZE - r.file_ofs);
+        hal.storage->read_block((uint8_t *)buf, r.file_ofs, count);
```
Comment (tridge) [inline libraries/AP_Filesystem/AP_Filesystem_Sys.cpp:170]: much simpler to just have a hal.storage function to return a const uint8_t * for storage. All the HALs have a memory backed copy

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_Filesystem/AP_Filesystem_Sys.cpp (hunk 1 above)
Comment (bugobliterator) [reply libraries/AP_Filesystem/AP_Filesystem_Sys.cpp:170]: yes, read_block does memcpy from that memory backed copy.

**Comment by tridge** [comment]
need to confirm the resulting file can be used with SITL

**Comment by bugobliterator** [comment]
> need to confirm the resulting file can be used with SITL

tested working

**Review by tridge — APPROVED** [review]

**Review by tridge — APPROVED** [review]

**Pull request merged**